feat(platforms)!: support wing platforms (#4669)
winglang/wing: docs/docs/02-concepts/03-platforms.md

Reviewer: These docs are great

---

> title: Platform
> id: platform
> description: Wing platforms
> keywords: [Wing platform, provisioning engines, multi-cloud]

Reviewer 2: ```suggestion
title: Platforms
id: platforms
description: Wing platforms
keywords: [platforms, targets, target, platform, aws, gcp, azure, sim, terraform, cloudformation, pulumi, provisioning engines, multi-cloud]
```

---

When working with the Wing programming language, an integral part of the compilation process is the use of platform. In essence, platform specify how and where your application is deployed. They determine both the cloud environment and the provisioning engine that the code will be deployed with.

## Platforms

You can view the list of available builtin platform with the `wing compile --help` command. Here is an example of the output:

```sh

> mywing compile --help         

Reviewer 2: ```suggestion
wing compile --help
```

Usage: wing compile [options] [entrypoint]

Compiles a Wing program

Arguments:
  entrypoint                    program .w entrypoint

Options:
  -t, --platform <platform...>  Target platform provider (builtin: sim, tf-aws, tf-azure, tf-gcp, awscdk) (default: ["sim"])
  -h, --help                    display help for command
```

> In the Options section above, there are several builtin platform that are shipped with Wing, such as `sim`, `tf-aws`, `tf-azure`, `tf-gcp`, and `awscdk`.

Reviewer 2: ```suggestion
Wing is shipped with several builtin platforms such as `sim`, `tf-aws`, `tf-azure`, `tf-gcp`, and `awscdk`, but it is also possible to create and use [custom platforms] to fully control how Wing resources are deployed to the cloud.
```

Add a link to [custom platforms] below.

These providers contain a combination of provision engine and cloud environment in their names, we refer to these as the platform models (which is discussed in more detail below). The only exception is `sim`, which is a special platform for testing and simulating applications locally.

> ### Specifying Multiple Platforms

Reviewer 2: I would put this below "custom platforms".

You may have noticed that the `--platform` option accepts multiple arguments. This means you can specify multiple platforms to compile your application to. For example, if you wanted to compile your application using multiple platforms

```sh

> wing compile app.main.w --platform tf-aws platform-foo platform-bar

Reviewer 2: Don't you need to specify `-p` multiple times in this case?

Author: Nope, commander variadic commands just allow you to keeep space separated. Which is kinda nice and kinda not because it makes it need to come after the entrypoint file. I think I’d prefer needing multiple -p but I haven’t found a way that works

Though not currently implemented, the platform model system is designed with extensibility in mind, as it will be used to determine compatibility between different platforms ([tracking issue](https://github.com/winglang/wing/issues/1474))

#### Model-specific code

There might be times when you need to write code that is specific to a particular platform model. For example, you may want to activate a verbose logging service only when testing locally to save on cloud log storage costs.

With the Wing `util` library, you can access environment variables. The `WING_MODEL` environment variable contains the current platform model as it's value, which you can use to conditionally run model-specific code. See the example below:

```js playground
bring cloud;
bring util;

let invocationCounter = new cloud.Counter();
let queue = new cloud.Queue();

queue.setConsumer(inflight (msg: str) => {
  invocationCounter.inc();
});

new cloud.Function(inflight ()=> { 
  // push a message to queue
  queue.push("m");
  // sleep according to model 

>   if util.env("WING_MODEL") == "sim" {

Reviewer 2: I am thinking that perhaps we should use the term "target" instead of "model". WDYT? It would work and might feel a bit more natural, especially in this context.

Author: I feel like there might be some more exploring that needs done here, I agree in this context "target" sounds better, then in others I feel like "model" is a more accurate term. That said, Im in favor of going back to target for now

    log("Running on Simulator, sleeping for 1s");
    util.sleep(1s);
  } else {
    log("Running on the cloud, sleeping for 30s");
    util.sleep(30s);
  }
  log("Function invoked ${invocationCounter.peek()} times");
});

```

## Custom Platforms

Wing's platform architecture is not just limited to the built-in platforms that come with the language; it's extensible. This means you can create custom platforms tailored to your needs, whether to support a unique cloud provider, introduce additional optimization layers, or integrate with specific enterprise systems.

### Why might you want to create a custom platform?

There are many reasons why you might want to create a custom platform. Here are a few examples:

> - Specialized Requirements: PErhaps your organization utilizes a particular cloud provider that isn't supported by Wing, or maybe you've built your own cloud infrastructure. Custom platforms allow you to tailor the deployment process precisely to your needs.

Reviewer 2: ```suggestion
- Custom infrastructure requirements: Wing's built-in platforms are opinionated about how various resources are implemented. With custom platforms, you can fully control the infrastructure configuration of each cloud resource based on the needs, policies and constraints of your team.
```

- Enhanced Security: Some organizations have stringent security requirements. With custom platforms, you can embed additional security checks, audits, or encryption layers to suit these needs.

- Optimizations: Your organization may have developed optimization strategies that can help reduce costs or improve performance when deploying applications. Integrating these strategies into a custom platform can make them a seamless part of the deployment process.

### Creating a custom platform

> Developing a custom platform requires understanding and adhering to the IPlatform interface. This ensures that your custom platform can be integrated smoothly with the Wing compilation process. 

Reviewer 2: ```suggestion
Developing a custom platform requires understanding and adhering to the `IPlatform` interface. This ensures that your custom platform can be integrated smoothly with the Wing compilation process. 
```
